pkg/descheduler/evictions/evictions.go

```diff
@@ -77,20 +261,155 @@ func NewPodEvictor(
+						klog.V(3).InfoS("Pod with eviction in background completed. Removing pod from the cache.", "pod", klog.KObj(newPod))
+						if err := erCache.deletePod(newPod); err != nil {
+							// If the deletion fails the cache may block eviction
+							klog.ErrorS(err, "Unable to delete updated pod from cache", "pod", newPod)
+						}
+						return
+					}
+					// Ignore any pod that does not have eviction in progress
+					if _, exists := newPod.Annotations[EvictionInProgressAnnotationKey]; !exists {
+						// In case EvictionInProgressAnnotationKey annotation is not present/removed
+						// it's unclear whether the eviction was restarted or terminated.
+						// If the eviction gets restarted the pod needs to be removed from the cache
+						// to allow re-triggering the eviction.
+						if _, exists := oldPod.Annotations[EvictionInProgressAnnotationKey]; !exists {
+							return
+						}
+						// the annotation was removed -> remove the pod from the cache to allow to
+						// request for eviction again. In case the eviction got restarted requesting
+						// the eviction again is expected to be a no-op. In case the eviction
+						// got terminated with no-retry, requesting a new eviction is a normal
+						// operation.
+						klog.V(3).InfoS("Eviction in background canceled (annotation removed). Removing pod from the cache.", "annotation", EvictionInProgressAnnotationKey, "pod", klog.KObj(newPod))
+						if err := erCache.deletePod(newPod); err != nil {
+							// If the deletion fails the cache may block eviction
+							klog.ErrorS(err, "Unable to delete updated pod from cache", "pod", newPod)
+							return
+						}
+						return
+					}
+					// Pick up the eviction in progress
+					if err := erCache.addPod(newPod); err != nil {
+						klog.ErrorS(err, "Unable to add pod to cache", "pod", newPod)
+						return
+					}
+				},
+				DeleteFunc: func(obj interface{}) {
+					var pod *v1.Pod
+					switch t := obj.(type) {
+					case *v1.Pod:
+						pod = t
+					case cache.DeletedFinalStateUnknown:
+						var ok bool
+						pod, ok = t.Obj.(*v1.Pod)
+						if !ok {
+							klog.ErrorS(nil, "Cannot convert to *v1.Pod", "obj", t.Obj)
+							return
+						}
+					default:
+						klog.ErrorS(nil, "Cannot convert to *v1.Pod", "obj", t)
+						return
+					}
+					// Ignore pod's that are not subject to an eviction in background
+					if _, exists := pod.Annotations[EvictionRequestAnnotationKey]; !exists {
+						return
+					}
+					klog.V(3).InfoS("Pod with eviction in background deleted/evicted. Removing pod from the cache.", "pod", klog.KObj(pod))
+					if err := erCache.deletePod(pod); err != nil {
+						klog.ErrorS(err, "Unable to delete pod from cache", "pod", pod)
+						return
+					}
+				},
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf("unable to register event handler for pod evictor: %v", err)
+		}
+
+		podEvictor.registeredHandlers = append(podEvictor.registeredHandlers, handlerRegistration)
+
+		go erCache.run(ctx)
```

Comment:
Can create the new goroutine and run this from outside in `RunDeschedulerStrategies`?

Reply:
Pod evictor is not mean to be a controller. Thus, no run method there. Currently, it's the most suitable place for placing the cache. Yet, eventually, this needs to go out some place else. Let's leave the run method as it is for now.